### DISCUSSION TITLE
Question Regarding Parameter "data" in Function [srv_g_xxxxxx] in [tm_g_xxxxxx]

### DISCUSSION THREAD

**Discussion by 3fresh** [insightsengineering/teal.osprey#256] [Q&A]
I hope you're doing well! I noticed that the "data" parameter is not explicitly defined or assigned a value within the function call. I'm curious to understand the rationale behind this design choice. Could you kindly provide some insight into why the "data" parameter is left undefined in the function call and how it is intended to be used by the users of your function?

If the "data" parameter originates from the input "data" of "teal::init", is it possible that the property of "data" could be a "tdata" instead of a "reactive"?

![image](https://github.com/insightsengineering/teal.osprey/assets/54909691/65dc08c0-80ad-42eb-bf90-658565b579fc)

**Comment by chlebowa** [comment]
Hi @3fresh,

I don't quite understand your question but I do know that the error you are experiencing results from your installation being out of date.
`teal` sends data to modules as a reactive expression containing a `teal_data` object but your modules expect a `tdata` object.
Please install your module packages from the `main` branch (or wait for the CRAN release, which is coming soon) and try again.

**Reply by 3fresh** [reply]
Hi @chlebowa,

Thank you for your response and for addressing the issue I encountered.

To provide further context, the problematic module originates from `teal.osprey::tm_g_swimlane`, or more precisely, from my modified version of tm_g_swimlane_modified.R.

Initially, I created a "TealData" object named "data" and ran the code with `teal::init(data=data, modules=modules(front_page(), tm_data_table(), tm_g_swimlane()))`. This execution encountered no issues in my Shiny application. However, upon renaming tm_g_swimlane to "tm_g_swimlane_modified," the problem arose when I ran `teal::init(data=data, modules=modules(front_page(), tm_data_table(), tm_g_swimlane_modified()))`.

Hence, my question is: Apart from modifying "tm_g_swimlane.R," are there any other files or dependencies I need to modify to ensure smooth integration when customizing your existing modules? 

For reference, the versions of packages I am currently using are as follows:
teal                              0.14.0
teal.code                     0.4.1
teal.data                      0.3.0
teal.moudles.general  0.8.16
teal.osprey                  0.1.16
teal.slice                      0.4.0
teal.transform             0.4.0
teal.wodgets               0.4.0

**Comment by chlebowa** [comment]
The `TealData` class is no longer supported so something is definitely out of date. If this is a new-ish app, I highly recommend installing `teal` from CRAN (this will install the rest of the family as well). 

I don't know what modifications you made to the module. If it runs after updating the core packages, good. If not, look up `?teal::as_tdata`.

If you have trouble creating your data in the new way, refer to these guides: https://github.com/insightsengineering/teal/discussions/945

If you have more trouble, let us know :+1:

**Reply by 3fresh** [reply]
I appreciate your suggestion. My app is currently built on an older version of teal (0.14), while teal.osprey relies on teal (>=0.15). However, thanks to your guidance, I was able to downgrade from "TealData" to "tdata" successfully. This temporary solution allows me to load custom teal.osprey modules in the older version of teal without any issues.

I will explore the possibility of upgrading my entire app to align with the latest teal version.

I eagerly anticipate the forthcoming CRAN release and the enhancements it will bring~

**Comment by chlebowa** [comment]
Excellent.
Note that a newer version of `teal` is now available on CRAN: https://cran.r-project.org/web/packages/teal/index.html
